src/odemis/gui/comp/overlay/shapes.py

```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
 
     @abstractmethod
     def refresh(self):
-        """Refresh the shape's points VA after copy or move_to operation."""
+        """Refresh the shape's points VA after copy, move_to or restore_state operation."""
+        pass
+
+    def undo(self):
+        """Undo the last action."""
+        if self.undo_stack:
+            state = self.undo_stack.pop()
+            self.redo_stack.append(state)
+            self.restore_state(state)
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
+                "No undo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    def redo(self):
+        """Redo the last undone action."""
+        if self.redo_stack:
+            state = self.redo_stack.pop()
+            self.undo_stack.append(state)
+            self.restore_state(state)
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
```

Comment:
"warning" means that the error log panel will turn orange. Just for Ctrl+Z or Ctrl+Y, that sounds a little over the top. I'd argue a "info" level is enough.